### DISCUSSION TITLE
prCommitsPerRunLimit is not working as expected with renovate v38

### DISCUSSION THREAD

**Discussion by ckychris** [renovatebot/renovate#30521] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

renovatebot/github-action@v40.1.5 renovate-version v38.17.1

### Please tell us more about your question or problem

Hello team,

I would like to first thank the maintainers of renovate to make updating dependencies less painful. Saving countless developer's wrists. 😄 

Back to the bug report. From July 29, our renovate instances starts creating an unlimited amount of updates each run with renovate version v38. In v37 and before, `prCommitsPerRunLimit` is set to 3 and covers new branches and auto merges, only 3 commits is pushed for each run. We self host renovate using`renovatebot/github-action@v40.1.5`. I can still see `DEBUG: Commits limit = 3` in the logs, but it's no longer working as expected.

I have tested several runs with v37, and it works as expected. I have attached the relevant parts of our config in the log section.

This config is crucial to us as we have hundreds of repositories. Some have more than 70 dependencies that need to be updated. Each auto-merge only merges one upgrade, and renovate will attempt to rebase all other upgrades after each auto-merge. We relied on limiting the number of commits each run makes, and the consistent order of updates will make sure the auto-merge is within the set limit for each run. This massively reduces our CICD costs. Imagine (70x3)x 30 minutes of CICD vs (70 + 70 + 69 + 68 .... + 3 + 2)x 30 minutes of CICD.

## Minimal reproduction

I tried to recreate this using the minimal reproduction template, but it uses a different renovate version `"renovateVersion": "37.440.7"` (I can update this when mend switch to v38). Which resulted in a different error. The repo has 2 branches opened.

https://github.com/ckychris/minimal-reproduction-template-prCommitsPerRunLimit
Part of the log on mend: `DEBUG: Commits limit = null`
```
DEBUG: Repository config (invalid)
{
  "fileName": "renovate.json"
  "config": {
    "prCommitsPerRunLimit": 1,
    "dependencyDashboard": true
  }
}
```
Results
```
DEBUG: syncBranchState() (branch="renovate/ts-node-10.x")
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState (branch="renovate/ts-node-10.x")
DEBUG: getBranchPr(renovate/ts-node-10.x) (branch="renovate/ts-node-10.x")
DEBUG: findPr(renovate/ts-node-10.x, undefined, open) (branch="renovate/ts-node-10.x")
DEBUG: findPr(renovate/ts-node-10.x, undefined, closed) (branch="renovate/ts-node-10.x")
DEBUG: branchExists=false (branch="renovate/ts-node-10.x")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/ts-node-10.x")
DEBUG: Check for closed PR because recreating closed PRs is disabled. (branch="renovate/ts-node-10.x")
DEBUG: findPr(renovate/ts-node-10.x, chore(deps): update dependency ts-node to v10, !open) (branch="renovate/ts-node-10.x")
DEBUG: prAlreadyExisted=false (branch="renovate/ts-node-10.x")
DEBUG: Checking schedule(schedule=at any time, tz=null, now=2024-08-01T08:33:15.463Z) (branch="renovate/ts-node-10.x")
DEBUG: No schedule defined (branch="renovate/ts-node-10.x")
DEBUG: Branch needs creating (branch="renovate/ts-node-10.x")
DEBUG: Using reuseExistingBranch: false (branch="renovate/ts-node-10.x")
DEBUG: Setting current branch to main (branch="renovate/ts-node-10.x")
DEBUG: latest commit (branch="renovate/ts-node-10.x")
{
  "branchName": "main"
  "latestCommitDate": "2024-08-01T18:32:50+10:00"
}

DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (branch="renovate/ts-node-10.x")
DEBUG: npm.updateDependency(): devDependencies.ts-node = ^10.0.0 (branch="renovate/ts-node-10.x")
DEBUG: Updating ts-node in package.json (branch="renovate/ts-node-10.x")
DEBUG: updateArtifacts for updatedPackageFiles (branch="renovate/ts-node-10.x")
DEBUG: Updated 1 package files (branch="renovate/ts-node-10.x")
DEBUG: Getting updated lock files (branch="renovate/ts-node-10.x")
DEBUG: Writing package.json files (branch="renovate/ts-node-10.x")
{
  "packageFiles": [
    "package.json"
  ]
}

DEBUG: Writing any updated package files (branch="renovate/ts-node-10.x")
DEBUG: Writing package.json (branch="renovate/ts-node-10.x")
DEBUG: Writing updated .npmrc file to .npmrc (branch="renovate/ts-node-10.x")
DEBUG: Generating pnpm-lock.yaml for . (branch="renovate/ts-node-10.x")
DEBUG: Spawning pnpm install to create pnpm-lock.yaml (branch="renovate/ts-node-10.x")
DEBUG: Found pnpm constraint in package.json packageManager: 8.15.9 (branch="renovate/ts-node-10.x")
DEBUG: Using node constraint ">=18.20.3" from package.json (branch="renovate/ts-node-10.x")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (branch="renovate/ts-node-10.x")
DEBUG: Using containerbase dynamic installs (branch="renovate/ts-node-10.x")
DEBUG: Resolved stable matching version (branch="renovate/ts-node-10.x")
{
  "toolName": "node"
  "constraint": ">=18.20.3"
  "resolvedVersion": "v20.16.0"
}

DEBUG: Executing command (branch="renovate/ts-node-10.x")
{
  "command": "install-tool node v20.16.0"
}

DEBUG: exec completed (branch="renovate/ts-node-10.x")
{
  "durationMs": 440
  "stdout": "v20.16.0[08:33:15.813] INFO (198): Installing tool node@20.16.0...\n\n[08:33:15.821] INFO (198): tool already installed\n    tool: \"node\"\n10.8.1\n0.28.2\n[08:33:15.971] INFO (198): Installed tool node in 158ms.\n"
  "stderr": ""
}

DEBUG: Executing command (branch="renovate/ts-node-10.x")
{
  "command": "install-tool pnpm 8.15.9"
}

DEBUG: exec completed (branch="renovate/ts-node-10.x")
{
  "durationMs": 432
  "stdout": "[08:33:16.241] INFO (234): Installing npm pnpm@8.15.9...\n[08:33:16.249] INFO (234): tool already installed\n    tool: \"pnpm\"\n8.15.9\n[08:33:16.461] INFO (234): Installed tool pnpm in 220ms.\n"
  "stderr": ""
}

DEBUG: Executing command (branch="renovate/ts-node-10.x")
{
  "command": "pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile"
}

DEBUG: exec completed (branch="renovate/ts-node-10.x")
{
  "durationMs": 1089
  "stdout": "Progress: resolved 1, reused 0, downloaded 0, added 0\nProgress: resolved 147, reused 125, downloaded 16, added 0, done\nDone in 1s\n"
  "stderr": ""
}

DEBUG: pnpm-lock.yaml needs updating (branch="renovate/ts-node-10.x")
DEBUG: Updated 1 lock files (branch="renovate/ts-node-10.x")
{
  "updatedArtifacts": [
    "pnpm-lock.yaml"
  ]
}

DEBUG: 2 file(s) to commit (branch="renovate/ts-node-10.x")
DEBUG: Preparing files for committing to branch renovate/ts-node-10.x (branch="renovate/ts-node-10.x")
DEBUG: git commit (branch="renovate/ts-node-10.x")
{
  "deletedFiles": []
  "ignoredFiles": []
  "result": {
    "author": null,
    "branch": "renovate/ts-node-10.x",
    "commit": "003a5e56045aba8511a2dffc16fd7d84e14352c5",
    "root": false,
    "summary": {
      "changes": 2,
      "insertions": 91,
      "deletions": 0
    }
  }
}

DEBUG: HEAD https://api.github.com/repos/ckychris/minimal-reproduction-template-prCommitsPerRunLimit/git/refs/heads/renovate/ts-node-10.x/ = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=78) (branch="renovate/ts-node-10.x")
DEBUG: HEAD https://api.github.com/repos/ckychris/minimal-reproduction-template-prCommitsPerRunLimit/git/refs/heads/renovate/ts-node-10.x = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=72) (branch="renovate/ts-node-10.x")
DEBUG: resetToCommit(a812b304aebf6f77ac0c218f0a8c73555ce70b9f) (branch="renovate/ts-node-10.x")
DEBUG: Fetching branch renovate/ts-node-10.x (branch="renovate/ts-node-10.x")
DEBUG: Setting current branch to main (branch="renovate/ts-node-10.x")
DEBUG: latest commit (branch="renovate/ts-node-10.x")
{
  "branchName": "main"
  "latestCommitDate": "2024-08-01T18:32:50+10:00"
}

INFO: Branch created (branch="renovate/ts-node-10.x")
{
  "commitSha": "5363fe9aa64763d4be496828da7b524f2a06cdfa"
}

DEBUG: Ensuring PR (branch="renovate/ts-node-10.x")
DEBUG: There are 0 errors and 0 warnings (branch="renovate/ts-node-10.x")
DEBUG: getBranchPr(renovate/ts-node-10.x) (branch="renovate/ts-node-10.x")
DEBUG: findPr(renovate/ts-node-10.x, undefined, open) (branch="renovate/ts-node-10.x")
DEBUG: findPr(renovate/ts-node-10.x, undefined, closed) (branch="renovate/ts-node-10.x")
DEBUG: getPrCache() (branch="renovate/ts-node-10.x")
DEBUG: Fetching changelog: https://github.com/TypeStrong/ts-node (9.1.1 -> 10.9.2) (branch="renovate/ts-node-10.x")
DEBUG: Creating PR (branch="renovate/ts-node-10.x")
{
  "prTitle": "chore(deps): update dependency ts-node to v10"
}

DEBUG: Creating PR (branch="renovate/ts-node-10.x")
{
  "title": "chore(deps): update dependency ts-node to v10"
  "head": "ckychris:renovate/ts-node-10.x"
  "base": "main"
  "draft": false
}

DEBUG: PR created (branch="renovate/ts-node-10.x")
{
  "pr": 2
  "draft": false
}

DEBUG: Adding labels '' to #2 (branch="renovate/ts-node-10.x")
INFO: PR created (branch="renovate/ts-node-10.x")
{
  "pr": 2
  "prTitle": "chore(deps): update dependency ts-node to v10"
}

DEBUG: addParticipants(pr=2) (branch="renovate/ts-node-10.x")
DEBUG: setPrCache() (branch="renovate/ts-node-10.x")
DEBUG: Created Pull Request #2 (branch="renovate/ts-node-10.x")
DEBUG: PR is not configured for automerge (branch="renovate/ts-node-10.x")
DEBUG: setBranchCommit() (branch="renovate/ts-node-10.x")
```

### Logs (if relevant)

2024-08-01T07:46:52.2687991Z        "migratedConfig": {
2024-08-01T07:46:52.2747970Z          "username": "*******",
2024-08-01T07:46:52.2753637Z          "gitAuthor": "*******",
2024-08-01T07:46:52.2758837Z          "gitIgnoredAuthors": ["*******"],
2024-08-01T07:46:52.2792495Z          "repositories": ["*******"],
2024-08-01T07:46:52.2797223Z          "baseBranches": ["renovate/rollup"],
2024-08-01T07:46:52.2801754Z          "prCommitsPerRunLimit": 3,
2024-08-01T07:46:52.2806047Z          "dependencyDashboard": true,
2024-08-01T07:46:52.2810319Z          "prConcurrentLimit": 0,
2024-08-01T07:46:52.2811965Z          "prHourlyLimit": 0,
2024-08-01T07:46:52.2812903Z          "timezone": "Australia/Melbourne",
2024-08-01T07:46:52.2813892Z          "printConfig": true,
2024-08-01T07:46:52.2815067Z          "recreateWhen": "always",
2024-08-01T07:46:52.2816239Z          "allowPostUpgradeCommandTemplating": true,
2024-08-01T07:46:52.2821661Z          "allowedPostUpgradeCommands": [
2024-08-01T07:46:52.2822991Z            "^(?:yarn|pnpm) renovate-post-upgrade",
2024-08-01T07:46:52.2824407Z            "^(?:bin|scripts)/renovate-post-upgrade"
2024-08-01T07:46:52.2826626Z          ],
2024-08-01T07:46:52.2831036Z          "semanticCommits": "enabled",
2024-08-01T07:46:52.2871765Z          "addLabels": ["renovate", "dependencies"],
2024-08-01T07:46:52.2872799Z          "branchPrefix": "renovate/update/",
2024-08-01T07:46:52.2873711Z          "npmrc": "***********",
2024-08-01T07:46:52.2874454Z          "npmrcMerge": true,
2024-08-01T07:46:52.2875158Z          "hostRules": [ // omitted auth details
2024-08-01T07:46:52.2888519Z          ],
2024-08-01T07:46:52.2889181Z          "rangeStrategy": "bump",
2024-08-01T07:46:52.2889958Z          "extends": [ // omitted custom package version control presets
2024-08-01T07:46:52.2896574Z          ],
2024-08-01T07:46:52.2897256Z          "vulnerabilityAlerts": {
2024-08-01T07:46:52.2898253Z            "addLabels": ["renovate", "dependencies", "security"],
2024-08-01T07:46:52.2899590Z            "enabled": true
2024-08-01T07:46:52.2900351Z          },
2024-08-01T07:46:52.2901047Z          "lockFileMaintenance": {
2024-08-01T07:46:52.2902195Z            "branchPrefix": "renovate/lockfile-maintenance/",
2024-08-01T07:46:52.2903186Z            "enabled": true
2024-08-01T07:46:52.2903899Z          },
2024-08-01T07:46:52.2904537Z          "onboarding": true,
2024-08-01T07:46:52.2905492Z          "onboardingPrTitle": "chore(deps): add renovate.json",
2024-08-01T07:46:52.2906703Z          "onboardingConfigFileName": "renovate.json5",
2024-08-01T07:46:52.2907717Z          "onboardingConfig": {
2024-08-01T07:46:52.2908934Z            "$schema": "https://docs.renovatebot.com/renovate-schema.json",
2024-08-01T07:46:52.2910042Z            "ignoreDeps": []
2024-08-01T07:46:52.2910726Z          }
2024-08-01T07:46:52.2911263Z        }

**Comment by rarkins** [comment]
prCommitsPerRunLimit is a self hosted option only which means it's not valid to put it in renovate.json. Hence it is not possible to reproduce in the hosted app.

**Reply by ckychris** [reply]
Thank you @rarkins for pointing this out. So having a minimal reproduction repo couldn't reproduce the error I hoped. 

I dug a little deeper, I looked at the limit file, and there is only a single call that will increase the commits count. `incLimitedValue('Commits')`, which is only called in 2 functions, `pushCommit` and `mergeBranch`. The respective functions that called them have debug messages printed.

`lib\workers\repository\update\branch\automerge.ts` `tryBranchAutomerge` will prints `Branch automerged`. This is the only place where [mergeBranch](https://github.com/renovatebot/renovate/blob/4b008bac298eb16668084857a4c402d6190ab1ae/lib/util/git/index.ts#L861-L879) is called.

[pushCommit](https://github.com/renovatebot/renovate/blob/4b008bac298eb16668084857a4c402d6190ab1ae/lib/util/git/index.ts#L1133-L1149) is another function that prints `Pushing refSpec` and `git push`

Taking the minimal reproduction run as a comparison to one of our known good runs:
- (Theory) Having 404 errors will lead to a code path change after `git commit`, it will call `resetToCommit` and `Fetching branch` instead
- I think this issue is not specific for v38, now the reproduction showing v37 can already trigger this error
- Is there any reason you can think of why renovate didn't call the `pushCommit` function?
- Maybe there is another function renovate can use to create a new branch, which doesn't call `incLimitedValue('Commits')` properly?

```
// Known good run with v37 in our repos, we are getting the correct commits limit is reached message
2024-08-01T08:10:51.9546278Z DEBUG: git commit  ....
2024-08-01T08:10:51.9640678Z DEBUG: Pushing refSpec renovate/update/...repo
2024-08-01T08:10:52.5878239Z DEBUG: git push (repository=, baseBranch=, branch=) ... results ....
2024-08-01T08:10:52.5904295Z DEBUG: Setting current branch to renovate/rollup ...
2024-08-01T08:10:52.6187661Z DEBUG: latest commit ...
2024-08-01T08:10:52.6189560Z        "branchName": "renovate/rollup",
2024-08-01T08:10:52.6190592Z        "latestCommitDate": "2024-07-31T12:14:09+10:00"
```

```
// Minimal reproduction full log with this update attached in the main thread.
DEBUG: git commit (branch="renovate/ts-node-10.x")
{.. object}

DEBUG: HEAD https://api.github.com/repos/ckychris/minimal-reproduction-template-prCommitsPerRunLimit/git/refs/heads/renovate/ts-node-10.x/ = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=78) (branch="renovate/ts-node-10.x")
DEBUG: HEAD https://api.github.com/repos/ckychris/minimal-reproduction-template-prCommitsPerRunLimit/git/refs/heads/renovate/ts-node-10.x = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=72) (branch="renovate/ts-node-10.x")
DEBUG: resetToCommit(a812b304aebf6f77ac0c218f0a8c73555ce70b9f) (branch="renovate/ts-node-10.x")
DEBUG: Fetching branch renovate/ts-node-10.x (branch="renovate/ts-node-10.x")
DEBUG: Setting current branch to main (branch="renovate/ts-node-10.x")
DEBUG: latest commit (branch="renovate/ts-node-10.x")
{
  "branchName": "main"
  "latestCommitDate": "2024-08-01T18:32:50+10:00"
}
INFO: Branch created (branch="renovate/ts-node-10.x")
{
  "commitSha": "5363fe9aa64763d4be496828da7b524f2a06cdfa"
}
```

**Reply by rarkins** [reply]
Perhaps it's because platformCommit is in use and that code path doesn't trigger the count increment?

**Reply by ckychris** [reply, marked as answer]
@rarkins Spot on, I searched for `platformCommit` and I can see that `git.commitFiles` calls `pushCommit`. And `commitFiles` calls `pushFiles` in the same file, after [calling the Github API](https://github.com/renovatebot/renovate/blob/1bf328f4b1f079c8a64f05448507683720058d44/lib/modules/platform/github/index.ts#L2044-L2054) it doesn't trigger the count increment.

```   
return platformCommit === 'enabled'
      ? commitFiles(commitConfig)
      : git.commitFiles(commitConfig);
```

I think `mergeBranch` is irrelevant as it's unchanged. So the bug fix here is just to add an `incLimitedValue('Commits')` right after the API calls. 

EDIT
I have made this contribution to renovate here: https://github.com/renovatebot/renovate/pull/30540
